### ISSUE TITLE
Unable to load large corpora into memory because PatternPointer length can't exceed 2^32 bytes (32 bit size descriptor)

### ISSUE THREAD

**Issue by proycon** [proycon/colibri-core#42]
Whilst fine in most situations, this doesn't work for IndexedCorpus which loads an entire corpus into one PatternPointer. This prevents loading very large corpora (continuation of #41):

```
Loading corpus data...
Loaded 307725534 sentences; corpussize (bytes) = 9157735203
ERROR: Pattern too long for pattern pointer [9157735203 bytes,explicit]
terminate called after throwing an instance of 'InternalError'
  what():  Colibri internal error
```

Simply setting the size descriptor to a 64 bit integer would waste too much memory in most other situation so isn't an option either. I think we need a more flexible solution through templating.

**Comment by wanthalf** [comment]
As for me, loading the whole corpus in a single file is rather a burden anyway. A three level index would be more useful (file, sentence, token). But that would probably require larger changes of the whole architecture, I am afraid. And adding a third number to the index wouldn't solve the problem, even if both the "file" and "sentence" pointers remain 32-bit - together they still make up 64 bits.

**Comment by proycon** [comment]
This current line of implementation is a rather major refactoring which will take some more time, so I'm gonna start again and try a more 'quick' fix, although that might come at increases memory cost in certain computations.

**Comment by proycon** [comment]
> As for me, loading the whole corpus in a single file is rather a burden anyway. A three level index would be more useful (file, sentence, token). But that would probably require larger changes of the whole architecture, I am afraid. 

Yeah, that would be another major refactoring. I don't think we can go in that direction. You'd have to solve that in some wrapper stage (mapping file,sentence to some kind of aggregated sentence)

> And adding a third number to the index wouldn't solve the problem, even if both the "file" and "sentence" pointers remain 32-bit - together they still make up 64 bits.

Indeed, that would come at an extra memory penalty so I don't want to go there.

**Comment by proycon** [comment]
Solved in v2.5.0 release

**Comment by proycon** [comment]
Last but not least, benchmarks of 32-bit size descriptors vs 64-bit size descriptors (in patterpointers) show a small increase in peak memory:

[benchmarks32.txt](https://github.com/proycon/colibri-core/files/2658097/benchmarks32.txt)
[benchmarks64.txt](https://github.com/proycon/colibri-core/files/2658098/benchmarks64.txt)
